swap/src/bitcoin/wallet.rs

```diff
@@ -392,39 +375,37 @@ impl Client {
         Ok(())
     }
 
-    fn drain_script_notifications(&mut self) -> Result<()> {
-        let script_history = &mut self.script_history;
-        let electrum = &self.electrum;
-
-        for (script, history) in script_history.iter_mut() {
-            if std::iter::from_fn(|| electrum.script_pop(script).transpose())
-                .last()
-                .transpose()
-                .map_err(|e| anyhow!("Failed to pop script notification: {:?}", e))?
-                .is_some()
-            {
-                let new_history = electrum
-                    .script_get_history(script)
-                    .map_err(|e| anyhow!("Failed to get to script history: {:?}", e))?;
-
-                let transactions = new_history
-                    .iter()
-                    .map(|h| h.tx_hash.to_string())
-                    .collect::<Vec<_>>();
-                tracing::trace!(
-                    ?transactions,
-                    "Got notification for changes to script {}",
-                    script
-                );
-
-                *history = new_history;
+    fn update_script_histories(&mut self) -> Result<()> {
+        let mut histories = self
+            .electrum
+            .batch_script_get_history(self.script_history.keys())
+            .map_err(|e| anyhow!("Failed to get script histories {:?}", e))?;
+
+        if histories.len() != self.script_history.keys().len() {
+            return Err(anyhow::Error::from(UnexpectedNumberOfScriptHistories {
```

Comment:
Unless you are planning to name this error type somewhere, e.g. by calling `anyhow::Error::downcast` you can save yourself some typing and just create the error messages directly using `anyhow!` or even `bail!` if you want to also get rid of the explicit `return`.